### ISSUE TITLE
Incorrect error when using generic, class and conditional typing (#43237 still exists)

### ISSUE THREAD

**Issue by pjannesen** [microsoft/TypeScript#49072]
# Bug Report

I reported bug #43237 about a year ago. The bug is closed as fixed but it still exists current version including nightly.

### 🕗 Version & Regression Information

The incorrect error is introduced in version 3.9 and exists in nightly.

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=4.6.2&ssl=1&ssc=1&pln=9&pc=22#code/MYGwhgzhAECyCeAlApsA9gJwCbQN4ChojoAHDASwDcwAXZUgCgEo9DiBffT-USGBAMrIaAHgAqQmtGQAPOgDss-JKkxYAfK2414JegBFaYAWgCuGYMnEpg0uckXKba6AB880ANoBrZPABcEDQU8gDmALr+YPLw0OyaALzQYjZ2CkpwKujY0AD8mZLWqOquKaie4dD+yTYVANz4QA

https://www.typescriptlang.org/play?ts=4.6.2#code/MYGwhgzhAECyCeAlApsA9gJwCbQN4ChojoAHDNAF1SpxIAoBKPQ4gX33f1EhgQGVkFADwAVARWjIAHlQB2WXklSYsAPmacK8EsmgARMBTB80AVwzBkolMEkzk8xTZXQAPnmgBtANbJ4ALggKDABLWQBzAF1-MFl4aFZ1AF5oERs7OQU4JXRsaAB+bPFrVFVXNNRPSOh-VJsqgG58IA

### 💻 Code 1

```typescript
export interface IFieldDef
{
    [name: string]: any;
}

export class Record<TRec extends IFieldDef>
{
    private a():TRec|null {
        return null;
    }
    protected b():TRec|null {
        return null;
    }
    public c() {
        return this.a();
    }
    public d() {
        return this.b();
    }
} 

export class Set<TSet extends Record<IFieldDef>>
{
    public item(): TSet|null {
        return null;
    }
}

export type DS<TRec extends Record<IFieldDef> | { [key:string]:any }> = TRec extends Record<IFieldDef> ? Set<TRec> : TRec[];
```

### 💻 Code 2

```typescript
class MyRecord {
    private p() {
    }
}

class MySet<TSet extends MyRecord> {
}

type DataSource<TRec extends MyRecord | { [key:string]:any }> = TRec extends MyRecord ? MySet<TRec>|TRec[] : TRec[];
```

### 🙁 Actual behavior

Error is reported

### 🙂 Expected behavior

No error

**Comment by RyanCavanaugh** [comment]
This is weird; it works for an ambient class but fails for a nonambient one
```ts
/*
declare class MyRecord {
    private p();
    f(): void;
}
*/
class MyRecord {
    private p() {} 
    f(): void {}
}

class MySet<TSet extends MyRecord> {
    ex!: TSet;
}

type DataSource<TRec extends MyRecord | { [key:string]:any }> = TRec extends MyRecord ? MySet<TRec> | TRec[] : TRec[];
```

**Comment by weswigham** [comment]
`this` types are the difference - since the ambient class has no method bodies and no explicit `this` types, the class has no associated `this`-type parameter, so every instance of the props is from the same declared type (and so they unify in the union+intersection in the conditional's constraint check, thus the private prop exists and matches the constraint). Meanwhile, the version-with-impls has an implicit `this` type parameter associated with it (since a `this` reference may appear in a method body that we don't know about), and so when the intersection `MyRecord & {[key: string]: any}` is formed, the `this` type of the `MyRecord` in that intersection changes, so the private `p` property gets a new symbol, making it _different_ from the `p` symbol in `MyRecord & MyRecord` (really just `MyRecord`), thus meaning it doesn't unify across both union members and so doesn't exist on the union.

I've been working on #44785 recently which tracks back to the same root cause. It's looking somewhat substantial to fix, since doing so seems to require better `this`-instantiation for unions (so private/protected props unify across union members), and looser compatibility requirements on the symbols compared for private/protected prop tracking (namely, requiring one symbol be a clone of the other, rather than exactly the same symbol).

**Comment by fatcerberus** [comment]
> Meanwhile, the version-with-impls has an implicit this type parameter associated with it (since a this reference may appear in a method body that we don't know about)

Wait, if this is true it almost completely invalidates the arguments given in the past that class methods don’t have an automatic `this: this` for perf reasons…

**Comment by fatcerberus** [comment]
No, I know that, but one of the main arguments for not doing `this: this` on methods automatically, IIRC, was that "it would make all classes generic" (due to the implicit type parameter for `this`) which would have major performance ramifications... but it seems like that's already the case based on what you've said already.

**Comment by RyanCavanaugh** [comment]
This is fixed by [#50328](https://github.com/microsoft/TypeScript/pull/50328). The nonambient private-member and protected `this` examples report TS2344 in 4.6.2 and 4.9.0-dev.20220920, but both compile cleanly from 4.9.0-dev.20220921, including 6.0.3 and the current native compiler.
